gunrock/app/mf/mf_test.cuh

```diff
@@ -0,0 +1,336 @@
+// ----------------------------------------------------------------
+// Gunrock -- Fast and Efficient GPU Graph Library
+// ----------------------------------------------------------------
+// This source code is distributed under the terms of LICENSE.TXT
+// in the root directory of this source distribution.
+// ----------------------------------------------------------------
+
+/**
+ * @fille
+ * sssp_test.cu
+ *
+ * @brief Test related functions for SSSP
+ */
+
+#pragma once
+
+// TODO: change to other includes, according to the problem
+#ifdef BOOST_FOUND
+    // Boost includes for CPU Dijkstra SSSP reference algorithms
+    #include <boost/config.hpp>
+    #include <boost/graph/graph_traits.hpp>
+    #include <boost/graph/adjacency_list.hpp>
+    #include <boost/graph/dijkstra_shortest_paths.hpp>
+    #include <boost/property_map/property_map.hpp>
+#else
+    #include <queue>
+    #include <vector>
+    #include <utility>
+#endif
+
+namespace gunrock {
+namespace app {
+namespace mf {
+
+/******************************************************************************
+ * Housekeeping Routines
+ ******************************************************************************/
+
+/**
+ * @brief Displays the SSSP result (i.e., distance from source)
+ * @tparam T Type of values to display
+ * @tparam SizeT Type of size counters
+ * @param[in] preds Search depth from the source for each node.
+ * @param[in] num_nodes Number of nodes in the graph.
+ */
+template<typename T, typename SizeT>
+void DisplaySolution(T *array, SizeT length)
+{
+    if (length > 40)
+        length = 40;
+
+    util::PrintMsg("[", true, false);
+    for (SizeT i = 0; i < length; ++i)
+    {
+        util::PrintMsg(std::to_string(i) + ":"
+            + std::to_string(array[i]) + " ", true, false);
+    }
+    util::PrintMsg("]");
+}
+
+/******************************************************************************
+ * Template Testing Routines
+ *****************************************************************************/
+
+/**
+ * @brief Simple CPU-based reference SSSP ranking implementations
+ * @tparam      GraphT        Type of the graph
+ * @tparam      ValueT        Type of the distances
+ * @param[in]   graph         Input graph
+ * @param[out]  distances     Computed distances from the source to each vertex
+ * @param[out]  preds         Computed predecessors for each vertex
+ * @param[in]   src           The source vertex
+ * @param[in]   quiet         Whether to print out anything to stdout
+ * @param[in]   mark_preds    Whether to compute predecessor info
+ * \return      double        Time taken for the SSSP
+ */
+template <
+    typename GraphT,
+    typename ValueT = typename GraphT::ValueT>
+double CPU_Reference(
```

Comment:
I think this is the next step for you, understanding how to add your CPU reference code (that you wrote a while ago). Incorporating this within the gunrock layer will allow us to use it for testing later and also help us write what goes inside of the advance and filter operator.